numpy/f2py/cfuncs.py

```diff
@@ -659,7 +660,7 @@
         if (*len == -1)
             *len = (PyArray_ITEMSIZE(arr))*PyArray_SIZE(arr);
         STRINGMALLOC(*str,*len);
-        STRINGCOPYN(*str,PyArray_DATA(arr),*len+1);
+        STRINGCOPYN(*str,PyArray_DATA(arr),*len);
```

Comment:
I think you need two different versions of `strncpy`; one for copying into a null-terminated buffer (line 688), and one for copying into a non-null-terminated buffer (this line).